packages/util/src/liquidityProxy.ts

```diff
@@ -0,0 +1,1207 @@
+  const reserve = FPNumber.fromCodecValue(payload.reserves.tbc[collateralAssetId]);
+  const price = tbcReferencePrice(collateralAssetId, payload);
+
+  return reserve.mul(price);
+};
+
+const mapCollateralizedFractionToPenalty = (fraction: number): number => {
+  if (fraction < 0.05) {
+    return 0.09;
+  } else if (fraction >= 0.05 && fraction < 0.1) {
+    return 0.06;
+  } else if (fraction >= 0.1 && fraction < 0.2) {
+    return 0.03;
+  } else if (fraction >= 0.2 && fraction < 0.3) {
+    return 0.01;
+  } else {
+    return 0;
+  }
+};
+
+const sellPenalty = (collateralAssetId: string, payload: QuotePayload): FPNumber => {
+  const idealReservesPrice = idealReservesReferencePrice(FPNumber.ZERO, payload);
+  const collateralReservesPrice = actualReservesReferencePrice(collateralAssetId, payload);
+
+  if (collateralReservesPrice.isZero()) {
+    throw new Error(`TBC: Not enough collateral reserves ${collateralAssetId}`);
+  }
+
+  const collateralizedFraction = safeDivide(collateralReservesPrice, idealReservesPrice);
+  const penalty = mapCollateralizedFractionToPenalty(collateralizedFraction.toNumber());
+
+  return new FPNumber(penalty);
+};
+
+const tbcSellPrice = (
+  collateralAssetId: string,
+  amount: FPNumber,
+  isDesiredInput: boolean,
+  payload: QuotePayload
+): FPNumber => {
+  const collateralSupply = FPNumber.fromCodecValue(payload.reserves.tbc[collateralAssetId]);
+  const xorPrice = tbcSellFunction(FPNumber.ZERO, payload);
+  const collateralPrice = tbcReferencePrice(collateralAssetId, payload);
+  const xorSupply = safeDivide(collateralSupply.mul(collateralPrice), xorPrice);
+
+  if (isDesiredInput) {
+    const outputCollateral = safeDivide(amount.mul(collateralSupply), xorSupply.add(amount));
+
+    if (FPNumber.isGreaterThan(outputCollateral, collateralSupply)) {
+      throw new Error(`TBC: Not enough collateral reserves ${collateralAssetId}`);
+    }
+
+    return outputCollateral;
+  } else {
+    if (FPNumber.isGreaterThan(amount, collateralSupply)) {
+      throw new Error(`TBC: Not enough collateral reserves ${collateralAssetId}`);
+    }
+
+    const outputXor = safeDivide(xorSupply.mul(amount), collateralSupply.sub(amount));
+
+    return outputXor;
+  }
+};
+
+const tbcBuyPrice = (
+  collateralAssetId: string,
+  amount: FPNumber,
+  isDesiredInput: boolean,
+  payload: QuotePayload
+): FPNumber => {
+  const currentState = tbcBuyFunction(FPNumber.ZERO, payload);
+  const collateralPrice = tbcReferencePrice(collateralAssetId, payload);
+
+  if (isDesiredInput) {
+    const collateralReferenceIn = collateralPrice.mul(amount);
+    const underPow = currentState.mul(PRICE_CHANGE_COEFF).mul(new FPNumber(2));
+    const underSqrt = underPow.mul(underPow).add(new FPNumber(8).mul(PRICE_CHANGE_COEFF).mul(collateralReferenceIn));
+    const xorOut = safeDivide(underSqrt.sqrt(), new FPNumber(2)).sub(PRICE_CHANGE_COEFF.mul(currentState));
+    return FPNumber.max(xorOut, FPNumber.ZERO);
```

Comment:
maybe just add a function inside this file for `FPNumber.max(something, FPNumber.ZERO)`?